### PR TITLE
S3IP sysfs specification and S3IP sysfs framework HLD

### PR DESCRIPTION
**Pull request by tianshangfei**
SONiC is designed to be portable to a variety of network devices. Many devices share the same ASIC platform and only differ in device-specific hardware components, such as PSUs, fan modules, and environment sensors. Currently, ODM vendor provides drivers to expose the device-specific hardware through sysfs, allowing SONiC to communicate with them, as described in the [Porting Guide](https://github.com/sonic-net/SONiC/wiki/Porting-Guide). However, many inefficient porting works are still required for SONiC developers due to different drivers from different devices.

The S3IP sysfs specification defines a unified interface to access peripheral hardware on devices from different vendors, making it easier for SONiC to support different devices and platforms.

| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-buildimage | [Two platforms supporting S3IP SYSFS (TCS8400, TCS9400)](https://github.com/sonic-net/sonic-buildimage/pull/12386) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The PSU driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12887) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The FAN driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12888) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The Transceiver driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12889) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The sensor driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12890) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The CPLD and FPGA driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12891) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The LED and watchdog driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12892) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The slot and switch_rootsysfs driver framework module complies with s3ip sysfs specification ](https://github.com/sonic-net/sonic-buildimage/pull/12893) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The user framework module complies with s3ip sysfs specification](https://github.com/sonic-net/sonic-buildimage/pull/12894) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The demo driver complies with s3ip sysfs specification](https://github.com/sonic-net/sonic-buildimage/pull/12895) | GitHub issue/pull request detail |
| sonic-buildimage | [S3IP sysfs framework - The build project of s3ip framework](https://github.com/sonic-net/sonic-buildimage/pull/12896) | GitHub issue/pull request detail |

### CODE REVIEW

**Comment by zhangyanzhao** [comment]
@pettershao want to be the reviewer of this HLD.

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_specification.md:75]: I suggest adding a `Power Sensor sysfs` section. This would provide the power in milliwatts from various power-sensing devices throughout the platform. Most on-board DC-DC converters (also called PWM or pulse-width modulator chips) provide voltage, current and calculated power. They often provide power In and Power out, allowing easy measurement of efficiency. It might be more reliable to read this directly rather than compute it from current * voltage, which might obtain readings at different times and result in invalid computation. It could also be more accurate because multiplying current * power in the controller results in compound loss of precision (rounded value * rounded value).

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_specification.md:108]: I think a better title would be `PSU Information sysfs`, since the data comprise many individual readings including power, current, voltage, inventory info, etc. all related to system-level PSUs. Changing the title would also more clearly differentiate between the proposed `Power Information sysfs` in the comment above.

**Review by chrispsommers — CHANGES_REQUESTED** [review]
Great proposal, please see comments.

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_specification.md:108]: Yes, PSU Information sysfs would be more suitable

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_specification.md:75]: Prior to this, the requirement to directly collect the Power sensor was not collected (because it can be obtained normally via Current *voltage), so it was not covered in the first phase.

This is a good point, we can probably add this sysfs definition in the next stage.

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:61]: I wonder if stronger language should be used to set expectations about how thoroughly vendors implement this specification. For example, a switch main board might have 10-20 sensors; language in this document might specify that "Vendors *SHOULD* provide software access to *every* sensor capable of being read. For any available sensor, vendors *SHALL* provide S3IP sysfs access."

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:69]: I feel the 2.0 platform APIs does the same thing, isn't it? These APIs are platform independent and platform vendors can implement them as they need i.e. 

1. Either read them directly from the SysFS created by device drivers, or 
2. Read them raw and then apply some filters/masks/pre-processing steps to get the final value

What is the reason/motivation behind the SysFS?
With S3IP, I see you are able to manage 1 but what about point#2?

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:157]: Vendors can have drivers which are common to various NOS. Isn't the S3IP framework mandating them to create SONiC specific drivers?

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (jeff-yin) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:38]: By my count, this is the third proposal to unify vendor device implementations for SONiC. Two others are:

1) PDDF: https://github.com/sonic-net/SONiC/blob/master/doc/platform/brcm_pdk_pddf.md
2) PIT: https://github.com/sonic-net/SONiC/pull/1014/files

How is this differentiated from the other two, and can this feature leverage anything from the other two?

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (jeff-yin) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:268]: Like I mentioned in the review for PIT:
https://github.com/sonic-net/SONiC/pull/1014/files#r932428427

> I am concerned that this will be yet another configuration file that platform vendors will have to generate and maintain when bringing platforms in to SONiC. Can the platform configuration be derived from already-existing platform configs?

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:61]: Ok, accept, I'll add it

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:161]: We consider the case of the common driver, so by the configuration file, the path of the common driver, soft link to S3IP sysfs directory, do not need to create SONiC specific drivers

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:272]: We plan to be two-in-one with the PDDF framework. but in two steps. The first step was to integrate the S3IP SYSFS framework into the Sonic community.
At present, there are many vendors in china that comply with S3IP framework. The S3IP SYSFS Framework is integrated into the SONiC community so that Chinese ODM vendors can more easily contribute their existing platforms to the community,  and other vendors can also have the opportunity to adapt their platforms according to S3IP specifications and expand their business opportunities in China. This is good for SONiC community ecology;

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_specification.md:109]: Recommend "Power Supply Unit" information for clarity, especially if you add a "Power Sensor" section in the future per https://github.com/sonic-net/SONiC/pull/1068#discussion_r952925474

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:299]: Ecological -> Ecosystem is probably what you meant.

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:73]: "specificationized" is not a real word, I suggest "well-specified"

**Review by chrispsommers — CHANGES_REQUESTED** [review]
Minor grammar suggestions, otherwise looks good.

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (chrispsommers) [reply doc/s3ip_sysfs/s3ip_sysfs_specification.md:109]: Regarding https://github.com/sonic-net/SONiC/pull/1068#discussion_r952925474 how do intend to track this? Could you add a placeholder in the spec?

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:73]: ok，accept

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:299]: ok，accept

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_specification.md:109]: ok，I will update this

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:298]: *** Requirements: (PDDF) The requirements are proposed from technical point of view aimed at platform driver and APIs development in SONiC. Only those SysFS which are required by SONiC platform APIs are exposed.

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:299]: EcoSystem: (PDDF) PDDF is a new framework and it is developed in the SONiC context. Some ODM platforms are already using PDDF. PDDF is an underlying framework which ODMs can use for faster development but it does not exposes any fixed SysFS nodes to the user.

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:300]: Development Mode: (PDDF) ODM vendors can use common PDDF kernel drivers and user space common platform APIs. Only some platform specific device data needs to be provided by the ODMs in the form of JSON files. Validation is via usual SONiC CLIs.

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:301]: Flexibility: (PDDF) PDDF can be used on the platforms which use I2C bus to communicate with the peripheral devices. Platforms which use BMC can also be brought up using PDDF. In future PDDF would be supported on platform using PCIE FPGA devices.

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:299]: Please add this comment in the PDDF column for the row "EcoSystem"

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:298]: Please add this comment in the PDDF column for the row "Requirements"

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:300]: Please add this comment in the PDDF column for the row "Development Mode"

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (FuzailBrcm) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md:301]: Please add this comment in the PDDF column for the row "Flexibility"

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: already updated

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: already updated

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: already updated

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: already updated

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: already updated

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:83]: spelling error, should be:
S3IP sysfs specification : [specification](/doc/s3ip_sysfs/s3ip_sysfs_specification.md)

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:299]: > complete a closed-loop of requirements. standards

Should the period be a comma?

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_hld.md:306]: ecology -> ecosystem

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_specification.md:140]: The paths in the table below say `/sys_switchtransceiver/` not `/sys_switch/transceiver/`. Which is correct?

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (chrispsommers) [inline doc/s3ip_sysfs/s3ip_sysfs_specification.md:149]: What is the purpose of "Development Light?"

**Review by chrispsommers — CHANGES_REQUESTED** [review]
Minor spelling/punctuation errors, otherwise looks good!

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: accepted

File: doc/s3ip_sysfs/s3ip_sysfs_hld.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_hld.md]: accepted

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_specification.md]: It was a translation error and has been revised

**Comment by tianshangfei** [comment]
> Minor spelling/punctuation errors, otherwise looks good!

I Receive notification  that "SONIC202211 branch will be forked by end of this month per schedule". 
I have submitted my code PR for some time, and haven't received any review comments at present. I hope the 202211 version supports this feature, please help speed up the review process, thank you

**Review by chrispsommers — DISMISSED** [review]
LGTM, thanks!

**Comment by tianshangfei** [comment]
I modified it once before, and found that some alarms were not introduced by our modification, and some alarms were considered to be false positives by tool inspection.
I will modify it again. 
Does LGTM have to be cleared before PR can be merged?

**Comment by chrispsommers** [comment]
> I modified it once before, and found that some alarms were not introduced by our modification, and some alarms were considered to be false positives by tool inspection.
I will modify it again.
Does LGTM have to be cleared before PR can be merged?

I'm not sure I fully understand your question. LGTM = "Looks good to me," sorry if my use of this acronym was ambiguous. I am OK with the PR.

GitHub is saying someone with write access needs to approve, in other words a repo maintainer.

**Comment by tianshangfei** [comment]
> > I modified it once before, and found that some alarms were not introduced by our modification, and some alarms were considered to be false positives by tool inspection.
> > I will modify it again.
> > Does LGTM have to be cleared before PR can be merged?
> 
> I'm not sure I fully understand your question. LGTM = "Looks good to me," sorry if my use of this acronym was ambiguous. I am OK with the PR.
> 
> GitHub is saying someone with write access needs to approve, in other words a repo maintainer.

OK，github's code review tool is also called LGTM （refer to LGTM.com）, So there was a misunderstanding.
How can I speed up the merge process？ thanks.

**Comment by tianshangfei** [comment]
@lguohan I have submitted the code PR for some time, and no further comments were received. I hope the 202211 version supports this feature, Please help to review it. Thanks

**Comment by zhangyanzhao** [comment]
@lguohan can you please check if this one can be merged? Thanks.

**Review by FuzailBrcm — APPROVED** [review]
Looks okay to me

**Comment by zhangyanzhao** [comment]
@lguohan  can you please help to check if this PR can be merged? Reviewers have approved and build passed. Thanks.

**Review by jeff-yin — APPROVED** [review]

**Comment by tianshangfei** [comment]
@lguohan please help to review this HLD

**Comment by zhangyanzhao** [comment]
@lguohan  would you please help to review and merge this PR? Thanks.

**Comment by zhangyanzhao** [comment]
@lguohan , all the code PRs are merged, can you please check if this PR can be merged as well? Thanks.

**Comment by zhangyanzhao** [comment]
@tianshangfei can you please rebase this PR? Seems like the info is out of sync with master branch

**Comment by tianshangfei** [comment]
> @tianshangfei can you please rebase this PR? Seems like the info is out of sync with master branch
Ok, PR is synchronized with master branch

**Comment by zhangyanzhao** [comment]
@lguohan  can you please help to review and merge this HLD PR? The code PRs have been merged.

File: doc/s3ip_sysfs/s3ip_sysfs_specification.md
Comment (FuzailBrcm) [inline doc/s3ip_sysfs/s3ip_sysfs_specification.md:100]: @tianshangfei 
Are you considering a Fantray as one fan unit? Or Fan[n] refers to individual fans inside each Fantray?

I have some confusion about the "Number of Fan Motors" field. If a Fantray is one fan unit, then "Front" and "Rear" fans can be marked as two motors which have different speed, max & min speeds, spped tolerance etc.

Comment (tianshangfei) [reply doc/s3ip_sysfs/s3ip_sysfs_specification.md:100]: The definitions of fans and motors are as follows: a device contains n fan units, and a fan unit contains 1-2 motors. Each motor may have a fault, so the speed, maximum and minimum speed, speed tolerance, etc. are collected by the motor. "Front" and "back" are used to identify the direction of the fan unit. F2B is the abbreviation for front-to-back airflow.

**Review by adyeung — APPROVED** [review]

**Pull request merged**
